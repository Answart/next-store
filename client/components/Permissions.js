import PropTypes from 'prop-types';
import { StyledPermissionsTable } from './styles/TableStyles';
import { UpdatePermissions } from './Buttons';
import { PERMISSIONS } from '../config';


class UserPermissions extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
      permissions: PropTypes.array.isRequired,
    }).isRequired
  };
  state = {
    permissions: [ ...this.props.user.permissions ]
  };
  handlePermissionChange = (e) => {
    const { checked, value } = e.target;
    let updatedPermissions = [ ...this.state.permissions ];

    if (checked) {
      updatedPermissions.push(value);
    } else {
      updatedPermissions = updatedPermissions.filter(permission => permission !== value);
    }

    this.setState({ permissions: updatedPermissions });
  };
  render() {
    const { user } = this.props;
    return (
      <tr key={`${user.id}-permissions`} id={`${user.id}-permissions`}>
        <td>
          {user.name}
        </td>

        <td>
          {user.email}
        </td>

        {PERMISSIONS.map(permission => (
          <td key={`${user.id}-permission-${permission}`}>
            <label htmlFor={`${user.id}-permission-${permission}`}>
              <input id={`${user.id}-permission-${permission}`}
                type="checkbox"
                checked={this.state.permissions.includes(permission)}
                value={permission}
                onChange={this.handlePermissionChange}
              />
            </label>
          </td>
        ))}

        <td>
          <UpdatePermissions
            permissions={this.state.permissions}
            userId={this.props.user.id}
          />
        </td>
      </tr>
    );
  }
};

const Permissions = ({ users }) => (
  <StyledPermissionsTable>
    <thead>
      <tr>
        <th>Name</th>

        <th>Email</th>

        {PERMISSIONS.map(permission => (
          <th key={permission}>
            {permission}
          </th>
        ))}

        <th>Action</th>
      </tr>
    </thead>

    <tbody>
      {!!users.length && users.map(user => (
        <UserPermissions key={`${user.id}-permissions`} user={user} />
      ))}
    </tbody>
  </StyledPermissionsTable>
);

Permissions.propTypes = {
  users: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    permissions: PropTypes.array.isRequired,
  })).isRequired,
};


export default Permissions;
