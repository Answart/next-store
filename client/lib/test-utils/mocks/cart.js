import {
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION
} from '../../../graphql'
import { fakeCartItem } from '../utils';
import { mockVariant } from './variant';

const mockCartItem = fakeCartItem();


const addToCartMutationMock = {
  request: { query: ADD_TO_CART_MUTATION, variables: { id: mockVariant.id } },
  result: {
    data: {
      addToCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};

const updateCartItemMutationMock = (quantity) => ({
  request: { query: UPDATE_CARTITEM_MUTATION, variables: { id: mockCartItem.id, quantity } },
  result: {
    data: {
      updateCartItem: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
        quantity,
        variant: { ...mockCartItem.variant }
      },
    },
  },
});

const removeFromCartMutationMock = {
  request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: mockCartItem.id } },
  result: {
    data: {
      removeFromCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};


export {
  mockCartItem,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
}
