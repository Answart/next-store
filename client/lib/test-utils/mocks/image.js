import { CREATE_IMAGE_MUTATION } from '../../../graphql'
import { fakeImage } from '../utils';

const mockImage = fakeImage();
const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
}


const createImageMutationMock = (overrides) => ({
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      ...overrides,
    },
  },
  result: {
    data: {
      createImage: {
        __typename: 'Image',
        id: mockImage.id,
      },
    },
  },
});

const createImageMutationErrorMock = {
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      cloudinary_id: '',
      name: '',
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};


export {
  mockImage, mockImageVariables,
  createImageMutationMock, createImageMutationErrorMock,
}
