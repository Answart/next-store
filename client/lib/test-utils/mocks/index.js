export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
