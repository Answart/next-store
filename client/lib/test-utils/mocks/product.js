import {
  PRODUCT_QUERY, SHOP_PRODUCTS_QUERY,
  CREATE_PRODUCT_MUTATION,
  UPDATE_PRODUCT_MUTATION,
  DELETE_PRODUCT_MUTATION,
} from '../../../graphql'
import { mockUser, mockProduct, mockImage, mockVariant, mockShopProductsVariables } from '../utils';


const productQueryMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage
          }
        }]
      },
    },
  },
};

const productQueryNoVariantMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: []
      },
    },
  },
};

const createProductMutationMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      createProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const createProductMutationErrorMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const updateProductMutationMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      updateProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const updateProductMutationErrorMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const deleteProductMutationMock = {
  request: { query: DELETE_PRODUCT_MUTATION, variables: { id: mockProduct.id } },
  result: {
    data: {
      deleteProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const shopProductsQueryNameEmptyMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { name: mockUser.name }},
  result: {
    data: {
      products: [],
    },
  },
};

const shopProductsQueryProductMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { ...mockShopProductsVariables }},
  result: {
    data: {
      products: [{
        ...mockProduct,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage,
          }
        }]
      }],
    },
  },
};


export {
  productQueryMock, productQueryNoVariantMock,
  shopProductsQueryNameEmptyMock, shopProductsQueryProductMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
}
