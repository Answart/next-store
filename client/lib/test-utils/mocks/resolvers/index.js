export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
} from './local';
export {
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
} from './user';
export {
  createImageMutationMock, createImageMutationErrorMock,
} from './image';
export {
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
} from './product';
export {
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock
} from './variant';
export {
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
} from './cartItem';
