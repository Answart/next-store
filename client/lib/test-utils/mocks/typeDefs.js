import {
  fakeUser, fakeImage, fakeProduct, fakeVariant, fakeCartItem, fakeOrder, fakeOrderItem,
} from '../utils';


const mockUser = fakeUser();
const mockUser2 = fakeUser();
const mockImage = fakeImage();
const mockProduct = fakeProduct();
const mockVariant = fakeVariant();
const mockCartItem = fakeCartItem();
const mockOrder = fakeOrder();
const mockOrderItem = fakeOrderItem();

mockOrder.items.push(mockOrderItem);

const mockUsers = [
  mockUser,
  {
    ...mockUser2,
    id: '87654321',
    permissions: ['USER'],
  }
];
const mockProducts = [{
  ...mockProduct,
  image: mockImage,
  user: {
    __typename: mockUser.__typename,
    id: mockUser.id,
    name: mockUser.name,
  },
  variants: [{
    ...mockVariant,
    product: {
      __typename: mockProduct.__typename,
      id: mockProduct.id,
      image: mockImage,
    }
  }]
}];

const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
};

const mockShopProductsVariables = {
  name: mockUser.name,
  orderBy: 'createdAt_DESC',
  skip: 0,
  first: 1
 };


export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrder, mockOrderItem,
};
