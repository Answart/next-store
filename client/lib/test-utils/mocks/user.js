import {
  CURRENT_USER_QUERY,
  SIGNOUT_MUTATION,
  REQUEST_PASSWORD_RESET_MUTATION,
} from '../../../graphql'
import { fakeUser } from '../utils';
import { mockCartItem } from './cart';

const mockUser = fakeUser();


const userQueryEmptyCartMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [],
      },
    },
  },
};

const userQueryNoUserMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: { me: null }
  },
};

const userQueryCartItemMock = overrides => ({
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [{
          ...mockCartItem,
          ...overrides
        }],
      },
    },
  },
});

const signoutMutationMock = {
  request: { query: SIGNOUT_MUTATION },
  result: {
    data: {
      signout: {
        __typename: 'Message',
        success: true,
        message: 'Goodbye!'
      },
    },
  },
};

const requestPasswordResetMutationMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'answart@sbcglobal.net' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: true,
        message: 'Your request has been sent! Check your email.',
      }
    },
  },
};

const requestPasswordResetMutationErrorMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'err@g.c' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: false,
        message: 'No such user found for email err@g.c',
      }
    },
  },
};


export {
  mockUser,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock
};