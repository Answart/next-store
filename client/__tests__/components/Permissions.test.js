import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import Permissions from '../../components/Permissions';
import { mockUsers } from '../../lib/test-utils/mocks';


describe('<Permissions />', () => {
  let wrapper, rows, firstRow, secondRow;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider mocks={[]}>
        <Permissions users={mockUsers} />
      </MockedProvider>
    );
    rows = wrapper.find('UserPermissions');
    firstRow = rows.at(0);
    secondRow = rows.at(1);
  });
  afterAll(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('matches with snapshot', async () => {
    wrapper.update();
    expect(toJSON(wrapper.find('Header'))).toMatchSnapshot();
  });

  it('renders all components properly', async () => {
    expect(wrapper.find('thead').length).toBe(1);
    expect(wrapper.find('thead').find('th').length).toBe(9);
    expect(wrapper.find('tbody').length).toBe(1);
    expect(wrapper.find('tbody').find('tr').length).toBe(2);
    expect(rows.length).toBe(2);
    expect(firstRow.find('tr').length).toBe(1);
    expect(firstRow.find('td').length).toBe(9);
    expect(firstRow.find('input').length).toBe(6);
    expect(firstRow.find('UpdatePermissions').length).toBe(1);
    expect(secondRow.find('tr').length).toBe(1);
    expect(secondRow.find('td').length).toBe(9);
    expect(secondRow.find('input').length).toBe(6);
    expect(secondRow.find('UpdatePermissions').length).toBe(1);
  });


  it('renders user rows properly', async () => {
    const firstMockUserInputAdmin = firstRow.find('input').at(0);
    const firstMockUserInputUser = firstRow.find('input').at(1);
    expect(firstMockUserInputAdmin.length).toBe(1);
    expect(firstMockUserInputAdmin.props().value).toBe('ADMIN');
    expect(firstMockUserInputAdmin.props().checked).toBe(true);
    expect(firstMockUserInputUser.length).toBe(1);
    expect(firstMockUserInputUser.props().value).toBe('USER');
    expect(firstMockUserInputUser.props().checked).toBe(false);
    expect(secondRow.find('tr').length).toBe(1);
    expect(secondRow.find('td').length).toBe(9);
    const secondMockUserInputAdmin = secondRow.find('input').at(0);
    const secondMockUserInputUser = secondRow.find('input').at(1);
    expect(secondMockUserInputAdmin.length).toBe(1);
    expect(secondMockUserInputAdmin.props().value).toBe('ADMIN');
    expect(secondMockUserInputAdmin.props().checked).toBe(false);
    expect(secondMockUserInputUser.length).toBe(1);
    expect(secondMockUserInputUser.props().value).toBe('USER');
    expect(secondMockUserInputUser.props().checked).toBe(true);
  });

  it('clicking a permission input calls handlePermissionChange properly', async () => {
    const firstRowAdminInput = firstRow.find('input').at(0);
    expect(firstRowAdminInput.length).toBe(1);
    expect(firstRowAdminInput.props().value).toBe('ADMIN');
    expect(firstRowAdminInput.props().checked).toBe(true);
    firstRowAdminInput.simulate('change', {target: {checked: false, value: 'ADMIN'}});
    await wait();
    wrapper.update();
    wrapper.find('UserPermissions').at(0).instance().forceUpdate();
    wrapper.update();
    const updatedInput = wrapper.find('UserPermissions').at(0).find('input').at(0);
    expect(updatedInput.props().value).toBe('ADMIN');
    expect(updatedInput.props().checked).toBe(false);
    updatedInput.simulate('change', {target: {checked: true, value: 'ADMIN'}});
    await wait();
    wrapper.update();
    wrapper.find('UserPermissions').at(0).instance().forceUpdate();
    wrapper.update();
    const updatedInput2 = wrapper.find('UserPermissions').at(0).find('input').at(0);
    expect(updatedInput2.props().value).toBe('ADMIN');
    expect(updatedInput2.props().checked).toBe(true);
  });
});
