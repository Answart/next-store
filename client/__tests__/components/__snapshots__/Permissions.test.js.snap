// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Permissions /> matches with snapshot 1`] = `null`;
